libs/knowledge-store/ragstack_knowledge_store/document_graph.py

```diff
@@ -0,0 +1,114 @@
+from __future__ import annotations
+
+import itertools
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, List, Optional, Union, Iterator, Tuple
+
+from langchain_core.callbacks import CallbackManagerForRetrieverRun
+from langchain_core.documents import Document
+from langchain_core.retrievers import BaseRetriever
+from pydantic import Field
+
+
+def _texts_to_documents(
+    texts: Iterator[str],
+    metadatas: Iterator[dict],
+) -> Iterator[Document]:
+    for (
+        text,
+        metadata,
+    ) in zip(texts, metadatas):
+        yield Document(
+            page_content=text,
+            metadata=metadata,
+        )
+
+
+class DocumentGraph(ABC):
+
+    def add_texts(
+        self,
+        texts: Iterable[str],
+        metadatas: Optional[Iterable[dict]] = None,
+        *,
+        ids: Optional[Iterable[str]] = None,
+        links: Iterable[Tuple[str, str]] = None,
+        **kwargs: Any,
+    ) -> List[str]:
+        metadatas_it = iter(metadatas) if metadatas else itertools.repeat({})
+        documents = _texts_to_documents(iter(texts), metadatas_it)
+        return self.add_documents(documents, ids=ids, links=links, **kwargs)
+
+    @abstractmethod
+    def add_documents(
+        self,
+        documents: Iterable[Document] = None,
+        *,
+        ids: Optional[Iterable[str]] = None,
+        links: Iterable[Tuple[str, str]] = None,
+        **kwargs: Any,
+    ) -> List[str]:
+        """Add documents to the graph"""
+
+    @abstractmethod
+    def traversing_retrieve(
```

Comment:
I don't know if it's better to have one uber function vs multiple functions...

Reply:
Yeah -- I find the `@overload` pattern a pain -- I think separate named functions can be nice. I guess my thoughts here are:

1. Multiple functions: Probably nicest from a documentation perspective. Easy to see the docs for a specific function, how different things work, etc. May lead to duplicity in implementation, but not horrible. Implementations can just have the abstract / unimplemented behavior, so not much burden.
2. One uber / overloaded function: Can use the `@overload` for types, but documentation gets messy. Main benefit is if we want to make it really clear "there are N ways to retrieve, and here are your choices". But that can be accomplished with references in the docs, so I think I prefer multiple functions...
3. One function with a bunch of hidden `**kwargs`. This is my least favorite. It's really hard to figure out which kwargs to use for different things, etc.

So -- I guess this comes back to separate methods. Thoughts on naming to show relationships? I'm wondering if it would be better to have `retrieve_traverse` (eg., use `retrieve_` as a prefix to group them?)

Reply:
Do we actually support different retrieval methods? It seems like the point of using this store is to do hybrid graph/vector retrieval, and between `k` and `depth` you can control how that retrieval behaves. Isn't `traversing_retrieve(k=3, depth-0) ~= retrieve(k=3)`?

Reply:
I mean, we plan to right?